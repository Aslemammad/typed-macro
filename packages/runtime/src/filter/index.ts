export { FilterOptions, FilterPattern, createFilter, Filter } from './filter'
