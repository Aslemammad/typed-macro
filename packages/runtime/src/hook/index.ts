export { createHookContainer, HookContainer, CombinedHooks } from './hook'
