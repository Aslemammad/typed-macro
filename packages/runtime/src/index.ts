export * from './transformer'
export * from './normalizer'
export * from './renderer'
export * from './runtime'
export * from './loader'
export * from './filter'
export * from './hook'
