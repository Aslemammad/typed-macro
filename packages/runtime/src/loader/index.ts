export { Loader, createLoader } from './loader'
