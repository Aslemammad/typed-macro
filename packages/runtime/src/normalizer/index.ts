export {
  ModularizedMacros,
  ModularizedModules,
  ModularizedTypes,
  NormalizedExports,
  normalizeExports,
  normalizeMacroProvider,
  NormalizedMacroProvider,
} from './normalizer'
