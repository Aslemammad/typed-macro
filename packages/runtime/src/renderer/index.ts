export { TypeRenderer, createTypeRenderer } from './renderer'
