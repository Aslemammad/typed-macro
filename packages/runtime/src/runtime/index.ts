export { Runtime, RuntimeOptions, createRuntime } from './runtime'
