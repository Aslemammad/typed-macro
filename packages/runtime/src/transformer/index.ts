export {
  Transformer,
  createTransformer,
  TransformerOptions,
} from './transformer'
