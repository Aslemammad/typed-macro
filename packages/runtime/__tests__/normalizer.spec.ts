import { normalizeMacroProvider } from '../src'
import {
  defineMacroProvider,
  EnvContext,
  MacroProvider,
} from '@typed-macro/core'
import { createMacroRemove } from '../../core/__tests__/macros'

describe('function normalizeMacroProvider()', () => {
  let provider: MacroProvider
  beforeEach(() => {
    provider = defineMacroProvider((env) => {
      return {
        id: 'test',
        exports: {
          a: {
            macros: env.ssr ? [createMacroRemove()] : [],
          },
          b: {
            code: 'export const value = 1',
            types: 'export const value: number',
          },
        },
      }
    })
  })

  it('should work', () => {
    expect(
      normalizeMacroProvider(provider, {
        host: 'test',
        packageManager: 'test',
        projectPath: [''],
        dev: true,
        ssr: false,
      })
    ).toMatchSnapshot()
    expect(
      normalizeMacroProvider(provider, {
        host: 'test',
        packageManager: 'test',
        projectPath: [''],
        dev: true,
        ssr: true,
      })
    ).toMatchSnapshot()
  })

  it('should reject invalid providers', () => {
    const env: EnvContext = {
      host: 'test',
      packageManager: 'test',
      projectPath: [''],
      dev: true,
      ssr: false,
    }

    expect(() => normalizeMacroProvider({} as any, env)).toThrowError()

    // version
    provider.version = 'unknown'
    expect(() => normalizeMacroProvider(provider, env)).toThrowError()
  })

  it('should reject invalid providers', () => {
    const env: EnvContext = {
      host: 'test',
      packageManager: 'test',
      projectPath: [''],
      dev: true,
      ssr: false,
    }

    provider = defineMacroProvider({
      id: 'test',
      exports: { a: { macros: ['' as any] } },
    })

    // invalid provider
    expect(() => normalizeMacroProvider({} as any, env)).toThrowError()

    // invalid macro
    expect(() => normalizeMacroProvider(provider, env)).toThrowError()

    // invalid macro version
    const m = createMacroRemove()
    m.version = 'unknown'
    provider = defineMacroProvider({
      id: 'test',
      exports: { a: { macros: [m] } },
    })
    expect(() => normalizeMacroProvider(provider, env)).toThrowError()
  })
})
