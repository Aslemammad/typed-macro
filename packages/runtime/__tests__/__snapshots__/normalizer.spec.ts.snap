// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function normalizeMacroProvider() should work 1`] = `
Object {
  "exports": Object {
    "a": Object {
      "macros": Array [],
    },
    "b": Object {
      "code": "export const value = 1",
      "types": "export const value: number",
    },
  },
  "id": "test",
  "normalizedExports": Object {
    "codes": Object {
      "b": "export const value = 1
",
    },
    "macros": Object {
      "a": Array [],
    },
    "types": Object {
      "a": Array [],
      "b": Array [
        "export const value: number",
      ],
    },
  },
}
`;

exports[`function normalizeMacroProvider() should work 2`] = `
Object {
  "exports": Object {
    "a": Object {
      "macros": Array [
        Object {
          "__is_macro": true,
          "handler": [Function],
          "name": "noop",
          "types": "export function noop(v: unknown): void",
          "version": "1",
        },
      ],
    },
    "b": Object {
      "code": "export const value = 1",
      "types": "export const value: number",
    },
  },
  "id": "test",
  "normalizedExports": Object {
    "codes": Object {
      "b": "export const value = 1
",
    },
    "macros": Object {
      "a": Array [
        Object {
          "__is_macro": true,
          "handler": [Function],
          "name": "noop",
          "types": "export function noop(v: unknown): void",
          "version": "1",
        },
      ],
    },
    "types": Object {
      "a": Array [
        "export function noop(v: unknown): void",
      ],
      "b": Array [
        "export const value: number",
      ],
    },
  },
}
`;
